ssl/t1_lib.c

```diff
@@ -2363,8 +2438,27 @@ int tls_choose_sigalg(SSL *s, int fatalerrs)
                 return 1;
 
         if (SSL_USE_SIGALGS(s)) {
+            size_t i;
+            /*
+             * Look for a certificate indicated by signature_algorithms_cert,
+             * if present.
+             */
+            if (s->s3->tmp.peer_cert_sigalgs != NULL) {
+                for (i = 0; i < s->s3->tmp.peer_cert_sigalgslen; i++) {
+                    lu = tls1_lookup_sigalg(s->s3->tmp.peer_cert_sigalgs[i]);
+                    if (lu == NULL || !ssl_has_cert(s, lu->sig_idx))
```

Comment:
Doesn't this check that the *keys* in the certificate are of the specified type, rather than checking that the *signatures* in the certificate are of the specified type

Reply:
Also we don't currently check the signatures for the chain at all at the moment. Should we?

Reply:
We probably should check signatures in the chain, yes.  (Though IIRC there is still a cop-out in the spec that lets you send anything if you can't comply with the peer's request, as a "better-than-nothing" sort of effort.)
And yes, this does check the key in the cert rather than the signature on the cert.  (I don't actually know how to check the latter off the top of my head.)
It seems like there is then a pretty complicated interaction between the signature_algorithms and signature_algorithms_cert processing, in that we have to pick a (leaf) cert that has a key that is compatible with the signature_algorithm that we'll negotiate.  Unless we want to claim that that's an error by the peer and it's okay to fail the connection when we try to perform an operation that can't work.

Reply:
> I don't actually know how to check the latter off the top of my head.

`X509_get_signature_info()`?